### PR TITLE
Improving 3D, fixing translateZ issues

### PR DESCRIPTION
**Pull request by jonastreub**
# Overview

Work on improving 3D started with the [originZ proposal](https://github.com/koenbok/Framer/pull/255) by Niels van Hoorn.

#### new properties on `Layer`
- `originZ`, improves rotational transform capabilities
- `flat`, all layers now preserve 3D unless flat is set to true
- `perspectiveOriginX` and `perspectiveOriginY`, both default to 0.5 which is the center
- `backfaceVisible`, allows the back of the layer to be hidden

#### new properties on `Context`
- `perspective`, defaults to 1200 so all layer transforms in the z direction are observable without any setup
- `backgroundColor`, since backgroundLayers cover layers positioned on the negative z axis it is recommended to set a backgroundColor on the Context
- `perspectiveOriginX` and `perspectiveOriginY`, both default to .5 which is the center

The described properties of context are available on the `Screen` object for easy adjustments

```coffee
Screen.perspective = 800 # default is 1200

Screen.backgroundColor = "white" # default is transparent
```

### Example

Using the new `originZ` property creating a carousel in 3D is quite easy. No more hard calculations.

```coffee
for index in [0...18]
	l = new Layer
		originZ: -300
		rotationY: 20 * index
		backgroundColor: Color.random(.5)
	l.center()
```
screen shot 2015-12-12 at 12 15 39

### CODE REVIEW

**Comment by koenbok** [comment]
Nice overview! Some questions I'd love to get answered before merging:

- What can you say about the performance impact of these?
- How does it impact Framer on desktop/mobile?
- What has been written about these properties in relation to performance?

**Comment by jonastreub** [comment]
The performance is pretty good though the fps drops a bit earlier both on desktop and mobile. I didn't notice a difference in performance trying out existing prototypes like the [Weather example](http://share.framerjs.com/bxlo8z7vrarz/).

perf

File: test/tests/LayerTest.coffee [hunk 1]
```diff
@@ -915,17 +939,17 @@ describe "Layer", ->
 			inputElement = _.first(inputElements)
 			inputElement.getAttribute("id").should.equal "hello"
 
-	describe "Force 2D", ->
+	# describe "Force 2D", ->
```
Comment (koenbok) [inline test/tests/LayerTest.coffee]: Why is this uncommented?

File: framer/Context.coffee [hunk 2]
```diff
@@ -50,6 +50,10 @@ class exports.Context extends BaseClass
 		@_parent = options.parent
 		@_name = options.name
 		
+		@perspective = 1200
```
Comment (koenbok) [inline framer/Context.coffee]: What happens if you chance these now. Nothing right? Are they supposed to be static?

File: framer/Context.coffee [hunk 3]
```diff
@@ -62,6 +66,8 @@ class exports.Context extends BaseClass
 		@resetTimers()
 		@resetIntervals()
 
+		@perspective = @perspective
```
Comment (koenbok) [inline framer/Context.coffee]: ???

File: framer/Context.coffee [hunk 4]
```diff
@@ -254,3 +260,44 @@ class exports.Context extends BaseClass
...
+			if _.isNumber(value)
+				@_perspective = value
+				@_element?.style["perspective"] = @_perspective
+				@_element?.style["webkitPerspective"] = @_perspective
+
+	_updatePerspective: ->
+		@_element?.style["webkitPerspectiveOrigin"] = "#{@perspectiveOriginX * 100}% #{@perspectiveOriginY * 100}%"
+
+	@define "perspectiveOriginX",
+		get: ->
```
Comment (koenbok) [inline framer/Context.coffee]: Maybe:

```
return @_perspectiveOriginX if _.isNumber(@_perspectiveOriginX)
return .5
```

File: framer/Context.coffee [hunk 5]
```diff
@@ -254,3 +260,44 @@ class exports.Context extends BaseClass
...
+				@_element?.style["backgroundColor"] = new Color value.toString()
+
+	@define "perspective",
+		get: ->
+			return @_perspective
+		set: (value) ->
+			if _.isNumber(value)
+				@_perspective = value
+				@_element?.style["perspective"] = @_perspective
+				@_element?.style["webkitPerspective"] = @_perspective
```
Comment (koenbok) [inline framer/Context.coffee]: Double? We don't do that anywhere.

If we're going to do css prefixing this would be my least favorite way :-)

**Comment by koenbok** [comment]
This is ready to merge in after our coming update.

**Comment by koenbok** [comment]
This looks done. Let's go.

**Pull request merged**